### PR TITLE
Add closingType and inReplyTo to various Q/A Messages

### PR DESCRIPTION
**Pull request by igrep**
The most important commit is https://github.com/iij-ii/direct-hs/commit/a3be5a3fe3bd2bac95f169c5da96bbd9173699d5
- Use `DuplicateRecordFields` for easier abstraction of record fields.
- Add `Generic` instances to `Message` and its related types to make them compatible with generic-lens.
- Plus revised version of https://github.com/iij-ii/direct-hs/pull/98/commits/965d775df409804333b4a3d40a3fad458ae7f93e

### CODE REVIEW

File: direct-hs/src/Web/Direct/Message.hs [hunk 1]
```diff
@@ -151,40 +162,61 @@ decodeMessage rspinfo = do
             M.ObjectWord 5   -> decodeFilesMessage
             M.ObjectWord 500 -> do
                 M.ObjectMap m <- look "content" rspinfo
-                qst           <- look "question" m >>= M.fromObject
-                return (YesNoQ qst)
+                YesNoQ <$> decodeYesNoQuestion m
             M.ObjectWord 501 -> do
                 M.ObjectMap m <- look "content" rspinfo
-                qst           <- look "question" m >>= M.fromObject
+                (YesNoQuestion qst ct) <- decodeYesNoQuestion m
```
Comment (kazu-yamamoto) [inline direct-hs/src/Web/Direct/Message.hs]: 丸括弧が不要な気がします。

File: direct-hs/src/Web/Direct/Message.hs [hunk 2]
```diff
@@ -51,64 +49,77 @@ encodeMessage (Files files mtext) tid =
...
 encodeMessage (Other text) tid =
     [M.ObjectWord tid, M.ObjectWord 1, M.ObjectStr text]
 
+encodeClosingType :: ClosingType -> (M.Object, M.Object)
+encodeClosingType ct = (M.ObjectStr "closing_type", M.ObjectWord w)
+  where
+    w = case ct of
+            OnlyOne -> 0
+            Anyone  -> 1
+
```
Comment (kazu-yamamoto) [inline direct-hs/src/Web/Direct/Message.hs]: ct の encoder / decoder を別途定義するとよいかも。

Comment (kazu-yamamoto) [reply direct-hs/src/Web/Direct/Message.hs]: `ClosingType` <-> `Int` という意味です。

**Review by yotsuya — APPROVED** [review]
LGTM

File: direct-hs/src/Web/Direct/Message.hs (hunk 2 above)
Comment (igrep) [reply direct-hs/src/Web/Direct/Message.hs]: こんな感じでいかがでしょうか！ https://github.com/iij-ii/direct-hs/pull/99/commits/f52bc3db929a2a7c512292d563a244b8b1c9c951

Comment (igrep) [reply direct-hs/src/Web/Direct/Message.hs]: こんな感じでいかがでしょうか！
https://github.com/iij-ii/direct-hs/pull/99/commits/f52bc3db929a2a7c512292d563a244b8b1c9c951

**Review by kazu-yamamoto — APPROVED** [review]
LGTM

**Pull request merged**